File: day_06/d06_p1.pl
#!/usr/bin/env perl

# Day 6: Part one

use strict;
use warnings;

use List::Util qw(reduce zip);
use POSIX qw(ceil floor);

sub get_nways {
    my ($t, $d) = @_;

    my $sq = sqrt($t * $t - 4 * $d);
    my $low = floor((($t - $sq) / 2) + 1);
    my $high = ceil((($t + $sq) / 2) - 1);

    return $high - $low + 1;
}

my @limit = split(/\s+/, <>);
shift @limit;

my @record = split(/\s+/, <>);
shift @record;

print((reduce { $a * $b } map { my ($t, $d) = @$_; get_nways($t, $d) } (zip \@limit, \@record)) . "\n");

File: day_06/d06_p1.clj
#!/usr/bin/env bb

;;;; # Day 6: Part one

(require '[clojure.java.io :as io])
(require '[clojure.math :as math])
(require '[clojure.string :as str])

(defn- number-of-win-ways
  [t d]
  (let [sq (math/sqrt (- (* t t) (* 4 d)))
        low (long (math/floor (inc (/ (- (double t) sq) 2.0))))
        high (long (math/ceil (dec (/ (+ (double t) sq) 2.0))))]
    (inc (- high low))))

(defn- make-race-info
  [line]
  (let [time-info (map parse-long (drop 1 (str/split (first line) #"\s+")))
        distance-info (map parse-long (drop 1 (str/split (second line) #"\s+")))]
    (zipmap time-info distance-info)))

(when (seq *command-line-args*)
  (->> (make-race-info (line-seq (io/reader (first *command-line-args*))))
       (map #(number-of-win-ways (first %) (second %)))
       (apply *)
       (println)))
